src/magic_folder/uploader.py

```diff
@@ -84,7 +84,7 @@ class LocalSnapshotCreator(object):
     _stash_dir = attr.ib(validator=attr.validators.instance_of(FilePath))
 
     @inline_callbacks
-    def store_local_snapshot(self, path):
+    def store_local_snapshot(self, path, relpath):
         """
```

Comment:
I don't really like this new API: it makes no connection between path and relpath; the caller could use "`FilePath('/etc/password')`" as the `path` and `"foo/bar"` as the `relpath` (which semantically doesn't make any sense). Since a newly-created snapshot must exist inside a particular magic-folder I think it makes more sense for the `LocalSnapshotCreator` to know its magic-folder path and keep the API the same (that is: pass in a single `FilePath`) and have this method compute the relpath piece. There will always be exactly one `LocalSnapshotCreator` for a particular magic-folder.

I see we check higher-up to report `ValueError`, but it still doesn't make sense for this piece to receive a `path` outside of its magic-directory. So basically all that boilds down to just moving the `segmentsFrom` call inside here so that the computed relpath is (still) definitely inside the magic-path.

Reply:
I disagree. I don't see any reason for them together. This is definitely coloured by thinking of this as a VCS, where interaction may be controlled by the user rather than automatic. Certainly we won't likely be doing anything like that for the MVP, but I think it is worthwhile keeping the distinction the in-process and stored representation (and the bits of code that interact with those) vs the on-disk representation (and the bits of code that want to interact with that).

Reply:
Partly, I'm thinking down the line when we are thinking about things like conflict resolution, we will want to deal with the remote state as distinct from the on-disk state.

One possible situation that we might run into is wanting to snapshot a path 'a/changed-file.local` with the name `a/changed-file` (without the `.local`), or maybe similarly a file that it outside the magic-folder in a temporary directory, but created by a merge tool.

Reply:
I think this makes things less-clear right now.

In the future, we can easily track relative-paths separately (if need be) but as it sits everything in the uploader is tracked by a single thing: the `FilePath` of the item. We have no use-case to do the relative-path computation "early" like this. Again, all I'm suggesting is moving the `segmentsFrom` call closer to where we need it.

Reply:
I'm not sure I entirely follow this change or the discussion.  The new `relpath` isn't documented - but from the implementation change I'm guessing it is meant to be a string giving the path to the changed file relative to the top of the magic-folder directory (whereas `path` is just the absolute path to the file in the system, since it is a `FilePath`).

So having these two distinct values means you can specify any file as the data source for any snapshot - in the magic-folder or not, with the same relative/basename or not.

I don't feel very strongly one way or the other.  If you want a tie-breaker, I'd say to go with the smaller interface for now (because the functionality of unrelated relpath and path values isn't used or exposed to users anywhere yet).  If the need arises later we can spend a bit more time thinking about what use-cases will really drive such a need and what the best design to support them is.

Reply:
We need the call where it is now, to be able to return an error message in a timely fashion, moving it would put the check behind the queue.